controllers/che/checluster_controller.go

```diff
@@ -260,6 +260,29 @@ func (r *CheClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
 		CheCluster: instance,
 	}
 
+	if isPreviousVersionDeployed(instance) {
+		// Current operator is newer than deployed Che
+		backupCR, err := getBackupCR(deployContext)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// Create a backup before updating current installation
+				if err := requestNewBackup(deployContext); err != nil {
+					return ctrl.Result{}, err
+				}
+				// Backup request is successfully submitted
+				// Give some time for the backup
+				return ctrl.Result{RequeueAfter: time.Second * 15}, nil
+			}
+			return ctrl.Result{}, err
+		}
+		if backupCR.Status.State == orgv1.STATE_IN_PROGRESS {
+			// Backup is still in progress
+			return ctrl.Result{RequeueAfter: time.Second * 5}, nil
```

Comment:
Maybe we could provide configurable backup timeout  to prevent hanging che-operator, or fail che-operator after this timeout...

Reply:
It is possible to add the timeout, but the logic for it will not be trivial as we cannot preserve a variable between reconcile loops. Actually, backup process has own timeouts, so it will notify main controller.